pkg/chunk/purger/purger.go

```diff
@@ -0,0 +1,576 @@
+	// and break the purge plan for other requests
+	inProcessRequestIDs    map[string]string
+	inProcessRequestIDsMtx sync.RWMutex
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// NewDataPurger creates a new DataPurger
+func NewDataPurger(cfg DataPurgerConfig, deleteStore *chunk.DeleteStore, chunkStore chunk.Store, storageClient chunk.ObjectClient) (*DataPurger, error) {
+	dataPurger := DataPurger{
+		cfg:                          cfg,
+		deleteStore:                  deleteStore,
+		chunkStore:                   chunkStore,
+		storageClient:                storageClient,
+		inProcessRequestIDs:          map[string]string{},
+		executePlansChan:             make(chan deleteRequestWithLogger, 50),
+		workerJobChan:                make(chan workerJob, 50),
+		workerJobExecutionStatusChan: make(chan workerJobExecutionStatus, 50),
+		quit:                         make(chan struct{}),
+	}
+
+	dataPurger.runWorkers()
+	go dataPurger.jobScheduler()
+
+	err := dataPurger.loadInprocessDeleteRequests()
+	if err != nil {
+		return nil, err
+	}
+
+	return &dataPurger, nil
+}
+
+// Run keeps pulling delete requests for planning
+func (dp *DataPurger) Run() {
+	dp.wg.Add(1)
+	defer dp.wg.Done()
+
+	pullDeleteRequestsToPlanDeletesTicker := time.NewTicker(time.Hour)
+	defer pullDeleteRequestsToPlanDeletesTicker.Stop()
+
+	for {
+		select {
+		case <-pullDeleteRequestsToPlanDeletesTicker.C:
+			err := dp.pullDeleteRequestsToPlanDeletes()
+			if err != nil {
+				level.Error(util.Logger).Log("msg", "error pulling delete requests for building plans", "err", err)
+			}
+		case <-dp.quit:
+			return
+		}
+	}
+}
+
+// Stop stops all background workers/loops
+func (dp *DataPurger) Stop() {
+	close(dp.quit)
+	dp.wg.Wait()
+}
+
+// we send all the delete plans to workerJobChan and then start checking for status on workerJobExecutionStatusChan
+func (dp *DataPurger) jobScheduler() {
+
+	pendingPlansCount := make(map[string]int)
+	pendingPlansCountMtx := sync.Mutex{}
+
+	dp.wg.Add(1)
+
+	go func() {
+
+		for {
+			select {
+			case status := <-dp.workerJobExecutionStatusChan:
+				if status.err != nil {
+					level.Error(status.logger).Log("msg", "error executing delete plan",
+						"plan_no", status.planNo, "err", status.err)
+					continue
+				}
+
+				err := dp.removeDeletePlan(context.Background(), status.userID, status.deleteRequestID, status.planNo)
+				if err != nil {
+					level.Error(status.logger).Log("msg", "error removing delete plan",
+						"plan_no", status.planNo, "err", err)
+				} else {
+					pendingPlansCountMtx.Lock()
+					pendingPlansCount[status.deleteRequestID]--
+
+					if pendingPlansCount[status.deleteRequestID] == 0 {
+						level.Info(status.logger).Log("msg", "finished execution of all plans, cleaning up and updating status of request")
+
+						err := dp.deleteStore.UpdateStatus(context.Background(), status.userID, status.deleteRequestID, chunk.Processed)
```

Comment:
I had it this way at start but because `DeleteStore` uses some of the components from `chunk` package and a future component that would need `DeleteStore` would be needed by `Store`s I felt it would be good to add it with other stores in `chunk` package. I think there will be a chance to move this around when I move that PR. Changing it now would mean a good amount of time in reviews again for you guys.